File: kdb-storage-engine/src/jvmTest/kotlin/dev/kdb/storage/engine/ReadThroughputBenchTest.kt
package dev.kdb.storage.engine

import dev.kdb.codec.KdbHash
import dev.kdb.codec.KdbUuid
import dev.kdb.document.KdbDocument
import dev.kdb.storage.StorageEngineConfig
import dev.kdb.storage.io.FileBackedPlatformIoShimFactory
import dev.kdb.storage.io.PlatformIoConfig
import dev.kdb.storage.wal.DefaultWriteAheadLogFactory
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.runBlocking
import kotlin.io.path.createTempDirectory
import kotlin.test.Test
import kotlin.time.measureTime

/**
 * Read-path counterpart to WriteThroughputBenchTest, requested alongside
 * the write numbers already in docs/benchmarks/phases-1-6-summary.md.
 * readBlob is a memTable lookup (no disk I/O once written); getDocument
 * is a ShardedDocStore lookup (Phase 2).
 */
class ReadThroughputBenchTest {
    @Test
    fun readBlobThroughput() =
        runBlocking {
            for (parallelism in listOf(1, 8, 64, 256)) {
                val root = createTempDirectory("kdb-bench-read-blob-kt").toString()
                val shim = FileBackedPlatformIoShimFactory.open(PlatformIoConfig(rootDirectory = root))
                val config = StorageEngineConfig(globalMemoryBudgetBytes = 64L shl 20, ioShim = shim)
                val wal = DefaultWriteAheadLogFactory().openOrCreate("bench-ns", config, shim)
                val engine = ServerStorageEngine("bench-ns", config, wal)

                val n = 5000
                val hashes = (0 until n).map { engine.writeBlob("payload-$it".encodeToByteArray()) }

                val opsPerWorker = 20000 / parallelism.coerceAtMost(20000)
                val totalOps = opsPerWorker * parallelism
                val elapsed =
                    measureTime {
                        coroutineScope {
                            (0 until parallelism).map {
                                async(Dispatchers.Default) {
                                    for (i in 0 until opsPerWorker) {
                                        engine.readBlob(hashes[i % n])
                                    }
                                }
                            }.forEach { it.await() }
                        }
                    }
                report("readBlob", parallelism, totalOps, elapsed)
            }
        }

    @Test
    fun getDocumentThroughput() =
        runBlocking {
            for (parallelism in listOf(1, 8, 64, 256)) {
                val root = createTempDirectory("kdb-bench-read-doc-kt").toString()
                val shim = FileBackedPlatformIoShimFactory.open(PlatformIoConfig(rootDirectory = root))
                val config = StorageEngineConfig(globalMemoryBudgetBytes = 64L shl 20, ioShim = shim)
                val wal = DefaultWriteAheadLogFactory().openOrCreate("bench-ns", config, shim)
                val engine = ServerStorageEngine("bench-ns", config, wal)

                val n = 5000
                val ids = (0 until n).map { KdbUuid.random() }
                ids.forEachIndexed { i, id -> engine.putDocument("bench-ns", KdbDocument(id, """{"v":$i}""")) }
                // getDocument ignores atCommit (see commitTree's doc comment - this
                // engine reflects live state, not per-branch history); pass any hash.
                val dummyCommit = KdbHash(ByteArray(32))

                val opsPerWorker = 20000 / parallelism.coerceAtMost(20000)
                val totalOps = opsPerWorker * parallelism
                val elapsed =
                    measureTime {
                        coroutineScope {
                            (0 until parallelism).map {
                                async(Dispatchers.Default) {
                                    for (i in 0 until opsPerWorker) {
                                        engine.getDocument("bench-ns", ids[i % n], dummyCommit)
                                    }
                                }
                            }.forEach { it.await() }
                        }
                    }
                report("getDocument", parallelism, totalOps, elapsed)
            }
        }

    private fun report(name: String, parallelism: Int, totalOps: Int, elapsed: kotlin.time.Duration) {
        val nsPerOp = elapsed.inWholeNanoseconds / totalOps.coerceAtLeast(1)
        val opsPerSec = if (nsPerOp == 0L) 0 else 1_000_000_000L / nsPerOp
        println("$name parallel=$parallelism totalOps=$totalOps elapsed=$elapsed nsPerOp=$nsPerOp opsPerSec=$opsPerSec")
    }
}

File: go/kdb/storage/engine/read_throughput_bench_test.go
package engine_test

// Read-path counterpart to write_throughput_bench_test.go, requested
// alongside the write numbers already captured in
// docs/benchmarks/phases-1-6-summary.md. Measures ReadBlob (memTable
// lookup, no disk I/O - blobs are served from memory once written) and
// GetDocument (sharded doc store lookup, Phase 2) under concurrency.

import (
	"fmt"
	"testing"

	"github.com/limidus/kdb/go/kdb/codec"
	"github.com/limidus/kdb/go/kdb/document"
)

func BenchmarkReadBlobConcurrent(b *testing.B) {
	for _, parallelism := range []int{1, 8, 64, 256} {
		b.Run(fmt.Sprintf("parallel-%d", parallelism), func(b *testing.B) {
			e := newDiskBackedServerEngine(b)
			const n = 5000
			hashes := make([]codec.Hash, n)
			for i := 0; i < n; i++ {
				h, err := e.WriteBlob([]byte(fmt.Sprintf("payload-%d", i)))
				if err != nil {
					b.Fatal(err)
				}
				hashes[i] = h
			}

			b.SetParallelism(parallelism)
			b.ReportAllocs()
			b.ResetTimer()
			b.RunParallel(func(pb *testing.PB) {
				i := 0
				for pb.Next() {
					h := hashes[i%n]
					if _, err := e.ReadBlob(h); err != nil {
						b.Fatal(err)
					}
					i++
				}
			})
		})
	}
}

func BenchmarkGetDocumentConcurrent(b *testing.B) {
	for _, parallelism := range []int{1, 8, 64, 256} {
		b.Run(fmt.Sprintf("parallel-%d", parallelism), func(b *testing.B) {
			e := newDiskBackedServerEngine(b)
			const n = 5000
			ids := make([]codec.UUID, n)
			for i := 0; i < n; i++ {
				id, err := codec.RandomUUID()
				if err != nil {
					b.Fatal(err)
				}
				doc, err := document.FromJSONWithID(id, fmt.Sprintf(`{"v":%d}`, i))
				if err != nil {
					b.Fatal(err)
				}
				if err := e.PutDocument("bench-ns", doc); err != nil {
					b.Fatal(err)
				}
				ids[i] = id
			}

			b.SetParallelism(parallelism)
			b.ReportAllocs()
			b.ResetTimer()
			b.RunParallel(func(pb *testing.PB) {
				i := 0
				for pb.Next() {
					id := ids[i%n]
					if _, err := e.GetDocument("bench-ns", id, codec.Hash{}); err != nil {
						b.Fatal(err)
					}
					i++
				}
			})
		})
	}
}
